### PR TITLE
[AutoDiff] Move thunk generation utilities to Utils/Differentiation/Thunk.h.

### PR DESCRIPTION
**Pull request by rxwei**
Move thunk generation utilities to a single place. This includes utilities for reabstraction thunks and subset parameters thunks.

Subset parameters thunk generators used to append new functions  `generatedFunctions` in `ADContext` for cleanup, but reabstraction thunk generators never did that so it seems unnecessary. I removed it so that AD thunk generation no longer depends on `ADContext` or `DifferentiationTransformer`.

### CODE REVIEW

File: lib/SILOptimizer/Mandatory/CMakeLists.txt [hunk 1]
```diff
@@ -7,6 +7,7 @@ silopt_register_sources(
   DefiniteInitialization.cpp
   # SWIFT_ENABLE_TENSORFLOW
   Differentiation.cpp
+  CanonicalizePartialApply.cpp
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/CMakeLists.txt]: Was this code addition unintentional?

**Review by dan-zheng — APPROVED** [review]

File: lib/SILOptimizer/Mandatory/CMakeLists.txt (hunk 1 above)
Comment (rxwei) [reply lib/SILOptimizer/Mandatory/CMakeLists.txt]: Thanks for the catch.

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

File: lib/SIL/SILVerifier.cpp [hunk 2]
```diff
@@ -1633,6 +1633,12 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
   }
 
   void checkPartialApplyInst(PartialApplyInst *PAI) {
+    if (PAI->getModule().getStage() != SILStage::Raw) {
+      require(!PAI->getFunctionType()->isDifferentiable(),
+              "partial_apply of differentiable funtions is only allowed "
```
Comment (dan-zheng) [inline lib/SIL/SILVerifier.cpp]: minor, in case you care to address later:
* `funtions` -> `functions`
* `SWIFT_ENABLE_TENSORFLOW`

Comment (rxwei) [reply lib/SIL/SILVerifier.cpp]: Darn it, this shouldn’t have been checked in.

**Pull request merged**